File: mobile/src/main/java/treehou/se/habit/ui/colorpicker/LightFragment.java
package treehou.se.habit.ui.colorpicker;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.gson.Gson;

import java.util.Locale;
import java.util.Timer;
import java.util.TimerTask;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.realm.Realm;
import se.treehou.ng.ohcommunicator.connector.models.OHServer;
import se.treehou.ng.ohcommunicator.connector.models.OHWidget;
import se.treehou.ng.ohcommunicator.services.IServerHandler;
import se.treehou.ng.ohcommunicator.util.GsonHelper;
import treehou.se.habit.R;
import treehou.se.habit.connector.Constants;
import treehou.se.habit.core.db.model.ServerDB;
import treehou.se.habit.module.HasActivitySubcomponentBuilders;
import treehou.se.habit.mvp.BaseDaggerFragment;
import treehou.se.habit.ui.colorpicker.LightContract.Presenter;
import treehou.se.habit.util.ConnectionFactory;

public class LightFragment extends BaseDaggerFragment<Presenter> implements LightContract.View {

    private static final String TAG = "LightFragment";

    private static final String ARG_SERVER = "ARG_SERVER";
    private static final String ARG_WIDGET = "ARG_SITEMAP";
    private static final String ARG_COLOR = "ARG_COLOR";

    @BindView(R.id.lbl_name) TextView lblName;
    @BindView(R.id.pcr_color_h) ColorPicker pcrColor;

    @Inject ConnectionFactory connectionFactory;
    @Inject Presenter presenter;

    private Realm realm;

    private OHServer server;
    private OHWidget widget;
    private int color;

    private Timer timer = new Timer();
    private Unbinder unbinder;

    public static LightFragment newInstance(long serverId, OHWidget widget, int color) {
        LightFragment fragment = new LightFragment();

        Bundle args = new Bundle();
        Gson gson = GsonHelper.createGsonBuilder();
        args.putLong(ARG_SERVER, serverId);
        args.putString(ARG_WIDGET, gson.toJson(widget));
        args.putInt(ARG_COLOR, color);
        fragment.setArguments(args);

        return fragment;
    }

    public LightFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        realm = Realm.getDefaultInstance();

        Bundle args = getArguments();
        long serverId = args.getLong(ARG_SERVER);
        String jWidget = args.getString(ARG_WIDGET);
        color = args.getInt(ARG_COLOR);

        Gson gson = GsonHelper.createGsonBuilder();
        server = ServerDB.load(realm, serverId).toGeneric();
        widget = gson.fromJson(jWidget, OHWidget.class);
    }

    @Override
    public Presenter getPresenter() {
        return presenter;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_colorpicker, container, false);
        unbinder = ButterKnife.bind(this, rootView);

        lblName.setText(widget.getLabel());
        pcrColor.setColor(color);

        return rootView;
    }

    @Override
    public void onResume() {
        super.onResume();
        pcrColor.setOnColorChangeListener(colorChangeListener);
    }

    @Override
    public void onPause() {
        super.onPause();
        pcrColor.setOnColorChangeListener(null);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        realm.close();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    private ColorPicker.ColorChangeListener colorChangeListener = new ColorPicker.ColorChangeListener() {
        @Override
        public void onColorChange(final float[] hsv) {
            timer.cancel();
            timer.purge();
            timer = new Timer();
            timer.schedule(new TimerTask() {

                @Override
                public void run() {
                    hsv[1] *= 100;
                    hsv[2] *= 100;

                    int hue = (int) hsv[0];
                    int saturation = (int) hsv[1];
                    int value = (int) hsv[2];

                    presenter.setHSV(widget.getItem(), hue, saturation, value);
                }
            }, 300);
        }
    };



    @Override
    protected void injectMembers(HasActivitySubcomponentBuilders hasActivitySubcomponentBuilders) {
        ((LightComponent.Builder) hasActivitySubcomponentBuilders.getFragmentComponentBuilder(LightFragment.class))
                .fragmentModule(new LightModule(this, getArguments()))
                .build().injectMembers(this);
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/colorpicker/LightFragment.kt
package treehou.se.habit.ui.colorpicker

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.realm.Realm
import kotlinx.android.synthetic.main.fragment_colorpicker.*
import se.treehou.ng.ohcommunicator.connector.models.OHServer
import se.treehou.ng.ohcommunicator.connector.models.OHWidget
import se.treehou.ng.ohcommunicator.util.GsonHelper
import treehou.se.habit.R
import treehou.se.habit.core.db.model.ServerDB
import treehou.se.habit.dagger.HasActivitySubcomponentBuilders
import treehou.se.habit.dagger.fragment.LightComponent
import treehou.se.habit.dagger.fragment.LightModule
import treehou.se.habit.mvp.BaseDaggerFragment
import treehou.se.habit.ui.colorpicker.LightContract.Presenter
import treehou.se.habit.util.ConnectionFactory
import java.util.*
import javax.inject.Inject

class LightFragment : BaseDaggerFragment<Presenter>(), LightContract.View {

    @Inject lateinit var connectionFactory: ConnectionFactory
    @Inject lateinit var lightPresenter: Presenter

    private var server: OHServer? = null
    private var widget: OHWidget? = null
    private var color: Int = 0

    private var timer = Timer()

    private val colorChangeListener = object : ColorPicker.ColorChangeListener {
        override fun onColorChange(hsv: FloatArray) {
            timer.cancel()
            timer.purge()
            timer = Timer()
            timer.schedule(object : TimerTask() {
                override fun run() {
                    hsv[1] = hsv[1] * 100f
                    hsv[2] = hsv[2] * 100f

                    val hue = hsv[0].toInt()
                    val saturation = hsv[1].toInt()
                    val value = hsv[2].toInt()

                    lightPresenter.setHSV(widget!!.item, hue, saturation, value)
                }
            }, 300)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        realm = Realm.getDefaultInstance()

        val args = arguments
        val serverId = args!!.getLong(ARG_SERVER)
        val jWidget = args.getString(ARG_WIDGET)
        color = args.getInt(ARG_COLOR)

        val gson = GsonHelper.createGsonBuilder()
        server = ServerDB.load(realm, serverId)?.toGeneric()
        widget = gson.fromJson(jWidget, OHWidget::class.java)
    }

    override fun getPresenter(): Presenter {
        return lightPresenter
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_colorpicker, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        nameLabel.text = widget!!.label
        pcrColor.color = color
    }

    override fun onResume() {
        super.onResume()
        pcrColor.setOnColorChangeListener(colorChangeListener)
    }

    override fun onPause() {
        super.onPause()
        pcrColor.setOnColorChangeListener(null)
    }

    override fun onDestroy() {
        super.onDestroy()
        realm.close()
    }

    override fun injectMembers(hasActivitySubcomponentBuilders: HasActivitySubcomponentBuilders) {
        (hasActivitySubcomponentBuilders.getFragmentComponentBuilder(LightFragment::class.java) as LightComponent.Builder)
                .fragmentModule(LightModule(this, arguments!!))
                .build().injectMembers(this)
    }

    companion object {

        private val TAG = "LightFragment"

        private val ARG_SERVER = "ARG_SERVER"
        private val ARG_WIDGET = "ARG_SITEMAP"
        private val ARG_COLOR = "ARG_COLOR"

        fun newInstance(serverId: Long, widget: OHWidget, color: Int): LightFragment {
            val fragment = LightFragment()

            val args = Bundle()
            val gson = GsonHelper.createGsonBuilder()
            args.putLong(ARG_SERVER, serverId)
            args.putString(ARG_WIDGET, gson.toJson(widget))
            args.putInt(ARG_COLOR, color)
            fragment.arguments = args

            return fragment
        }
    }
}
